packages/api/amplify_api/lib/src/graphql/ws/web_socket_connection.dart

```diff
@@ -0,0 +1,210 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:amplify_api/src/decorators/web_socket_auth_utils.dart';
+import 'package:amplify_core/amplify_core.dart';
+import 'package:async/async.dart';
+import 'package:meta/meta.dart';
+import 'package:web_socket_channel/web_socket_channel.dart';
+
+import 'web_socket_message_stream_transformer.dart';
+import 'web_socket_types.dart';
+
+/// {@template amplify_api.web_socket_connection}
+/// Manages connection with an AppSync backend and subscription routing.
+/// {@endtemplate}
+@internal
+class WebSocketConnection implements Closeable {
+  /// Allowed protocols for this connection.
+  static const webSocketProtocols = ['graphql-ws'];
+
+  // Config and auth repo together determine how to authorize connection URLs
+  // and subscription registration messages.
+  final AmplifyAuthProviderRepository _authProviderRepo;
+  final AWSApiConfig _config;
+
+  // Manages all incoming messages from server. Primarily handles messages related
+  // to the entire connection. E.g. connection_ack, connection_error, ka, error.
+  // Other events (for single subscriptions) rebroadcast to _rebroadcastController.
+  late final WebSocketChannel _channel;
+  late final StreamSubscription<WebSocketMessage> _subscription;
+  late final RestartableTimer _timeoutTimer;
```

Comment:
nit: Does this timer need to be cleaned up on close/cancel? What happens when connection ends, but time remains and then ends?

Reply:
Good call! For this PR, I think I will call cancel on the timeout when closing. I think with reconnect, we may need to reconsider some of this, maybe making some of these not final and re-assigning. Same thing for futures used during initial connection.